core/include/kernel/dt.h

```diff
@@ -0,0 +1,156 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef KERNEL_DT_H
+#define KERNEL_DT_H
+
+#include <compiler.h>
+#include <stdint.h>
+#include <types_ext.h>
+
+/*
+ * Bitfield to reflect status and secure-status values ("ok", "disabled" or not
+ * present)
+ */
+#define DT_STATUS_DISABLED 0
+#define DT_STATUS_OK_NSEC  1
+#define DT_STATUS_OK_SEC   2
+
+#if defined(CFG_DT)
+
+/*
+ * DT-aware drivers
+ */
+
+struct dt_device_match {
+	const char *compatible;
+};
+
+struct dt_driver {
+	const char *name;
+	const struct dt_device_match *match_table; /* null-terminated */
+	const void *driver;
+};
+
+#define __dt_driver __attribute__((__section__(".rodata.dtdrv")))
+
+extern struct dt_driver __rodata_dtdrv_start, __rodata_dtdrv_end;
+
+const struct dt_driver *dt_find_driver(const char *compatible);
+const struct dt_driver *__dt_driver_start(void);
+const struct dt_driver *__dt_driver_end(void);
+
+/*
+ * Map a device into secure or non-secure memory and return the base VA and
+ * the mapping size. The mapping is done with type MEM_AREA_IO_SEC or
+ * MEM_AREA_IO_NSEC, depending on the device status.
+ * If the mapping already exists, the function simply returns the @vbase and
+ * @size information.
+ *
+ * @offs is the offset of the node that describes the device in @fdt.
+ * @base receives the base virtual address corresponding to the base physical
+ * address of the "reg" property
+ * @size receives the size of the mapping
+ *
+ * Returns 0 on success or -1 in case of error.
+ */
+int dt_map_dev(const void *fdt, int offs, vaddr_t *base, size_t *size);
+
+/*
+ * FDT manipulation functions, not provided by <libfdt.h>
+ */
+
+/*
+ * Return the base address for the "reg" property of the specified node or
+ * (paddr_t)-1 in case of error
+ */
+paddr_t _fdt_reg_base_address(const void *fdt, int offs);
+
+/*
+ * Return the reg size for the reg property of the specified node or -1 in case
+ * of error
+ */
+ssize_t _fdt_reg_size(const void *fdt, int offs);
+
+/*
+ * Read the status and secure-status properties into a bitfield.
+ * @status is set to DT_STATUS_DISABLED or a combination of DT_STATUS_OK_NSEC
+ * and DT_STATUS_OK_SEC
+ * Returns 0 on success or -1 in case of error.
+ */
+int _fdt_get_status(const void *fdt, int offs, int *status);
+
+#else /* !CFG_DT */
+
+static inline const struct dt_driver *dt_find_driver(const char *compatible
+						     __unused)
+{
+	return NULL;
+}
+
+static inline const struct dt_driver *__dt_driver_start(void)
+{
+	return NULL;
+}
+
+static inline const struct dt_driver *__dt_driver_end(void)
+{
+	return NULL;
+}
+
+#define __dt_driver
```

Comment:
If `CFG_DT` is not defined, i could be nice to not define neither `__dt_driver` to insure one cannot implement DT dependent resources without enabling `CFG_DT=y`.